### ISSUE TITLE
Is it possible to implement the functionality of dragula in react-dnd?

### ISSUE THREAD

**Issue by igorpreston** [react-dnd/react-dnd#489]
Hi,

I've been using react-dnd for a while and I'd like to know if it is possible to reimplement [dragula-like](https://bevacqua.github.io/dragula/) functionality in react-dnd?

I want the next features:
1. Rendering preview container for dragged element where it's going to be dropped.
2. Sorting.
3. Cloning with drag'n'drop.

Any help is greatly appreciated! Thanks!

**Comment by igorpreston** [comment]
Hey, @gaearon sorry for mentioning you in the thread, but I thought your advise might help a lot of at least direct toward the right direction on how those features can be implemented manually. Thanks.

**Comment by raphaelkross** [comment]
Hey @igorpreston 

These features can be definitely implemented. Of course, things would be a bit more manual - I recommend the Sortable Example - https://gaearon.github.io/react-dnd/examples-sortable-simple.html

The things you would face a bit of problem will likely be:

- The preview is actually another element - there's no such thing like moving an object in DnD, you receive info when you start dragging and when dropping - this info can be the component itself (and you can render it normally in the parent, for example). Here, the isOver event would be useful.

- To know if you should render above or below, the Sortable presents an example of how this can be done

- The clone operation follows the first item here, basically you copy the infos when drag start and paste when done

...

In my case, I used it together with Redux - you can use any method that turns your drop areas in controlled components - which means they just render your components from an array (coming from your store, for example).

When an user start hovering the drop area, you can add the "preview" in the store (marked as preview) - if the drop doesn't happen, you delete - else, you add it - removing the "preview" status and removing the source one from the array.

All the operations happen in the array/store and your components just reflect the changes - I've found this to be the more robust and easy to work solution - in my case, at least!

**Comment by bluespore** [comment]
+1

I've got the same requirement right now to clone the source for the target. We have a list of widgets in a sidebar that we're using for a page builder of sorts. Each widget may or may not be used many times (determined by a prop, obvs), so I'd like to replicate the `copy` functionality from dragula. 

If anyone has an example of that, it'd be greatly appreciated. If I happen to tackle it before then, I'll post my example back here.

**Comment by davidspiess** [comment]
@bluespore may i ask if you already solved it?

**Comment by bluespore** [comment]
@davidspiess I can't remember exactly what I was attempting to solve back then, but looking at Dragula's "copy" function suggests that we managed to implement something similar (see attached)

![dragula-copy](https://cloud.githubusercontent.com/assets/2460681/25600291/5fe7a1ec-2f26-11e7-89a2-50140e9b8feb.gif)

Difficult to say exactly what will be beneficial to you as I'm fly-in fly-out on this comment but I'll try to summarise:

- We have a custom drag layer in a global container which switch cases `itemType` when a `DragSource` has begun dragging
- That switch case will decide what component to render beneath the mouse and pass props appropriately from the `monitor`
- There are specific `DropZone` components that are set up as `DropTarget`s, that are rendered based on some mid-loop logic for each item in the array that is rendered to the business-area of the application. This means, we have a set of logic that runs to determine if there's a `DropZone` rendered above / below the iterated item on that render. This is obviously where we can drop the dragged item.

The `source` property for our `DragSource` literally just returns all the props:
```
{
	beginDrag(props) {
		return { ...props }
	},
}
```

Hopefully that helps? Sorry, I'm digging at old code and not entirely sure where the difficulty lay when I first looked into it :)
